app/scripts/migrations/135.ts

```diff
@@ -0,0 +1,147 @@
+import { hasProperty, isObject } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 135;
+
+/**
+ * Removes tokens with `decimals === null` from `allTokens`, `allIgnoredTokens`, and `allDetectedTokens`.
+ * Captures exceptions for invalid states using Sentry and logs tokens with `decimals === null`.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly
+ * what we persist to disk.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  transformState(versionedData.data);
+  return versionedData;
+}
+
+/**
+ * Transforms the TokensController state to remove tokens with `decimals === null`.
+ *
+ * @param state - The persisted MetaMask state.
+ */
+function transformState(state: Record<string, unknown>): void {
+  if (!hasProperty(state, 'TokensController')) {
+    return;
+  }
+
+  const tokensControllerState = state.TokensController;
+
+  if (!isObject(tokensControllerState)) {
+    global.sentry?.captureException(
+      new Error(
+        `Migration ${version}: Invalid TokensController state of type '${typeof tokensControllerState}'`,
+      ),
+    );
+    return;
+  }
+
+  // Validate and transform `allTokens`
+  if (hasProperty(tokensControllerState, 'allTokens')) {
+    if (!isObject(tokensControllerState.allTokens)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid allTokens state of type '${typeof tokensControllerState.allTokens}'`,
+        ),
+      );
+    } else {
+      tokensControllerState.allTokens = transformTokenCollection(
+        tokensControllerState.allTokens,
+        'allTokens',
+      );
+    }
+  }
+
+  // Validate and transform `allIgnoredTokens`
+  if (hasProperty(tokensControllerState, 'allIgnoredTokens')) {
+    if (!isObject(tokensControllerState.allIgnoredTokens)) {
+      global.sentry?.captureException(
+        new Error(
+          `Migration ${version}: Invalid allIgnoredTokens state of type '${typeof tokensControllerState.allIgnoredTokens}'`,
+        ),
+      );
+    } else {
+      tokensControllerState.allIgnoredTokens = transformTokenCollection(
```

Comment:
Doesn't `allIgnoredTokens` have a different type than `allTokens` and `allDetectedTokens`?

Unlike the other 2 that map to an object, the types indicate `allIgnoredTokens` maps to just a string of the contract address:

https://github.com/MetaMask/core/blob/3ba11472793bce8e1b081be2f60cbcd75a394f08/packages/assets-controllers/src/TokensController.ts#L91

Reply:
e.g.

```json
    "allIgnoredTokens": {
      "0x1": {
        "0x31508a659fdd1e6736b3d92e3fd1a577a39cf304": [
          "0x3007083EAA95497cD6B2b809fB97B6A30bdF53D3"
        ]
      }
    },
```

Reply:
This may mean it doesn't need to be migrated, if it doesn't have any decimal properties

Reply:
yes indeed, that's what I saw after the push while doing manual tests :
```
export type TokensControllerState = {
  tokens: Token[];
  ignoredTokens: string[];
  detectedTokens: Token[];
  allTokens: { [chainId: Hex]: { [key: string]: Token[] } };
  allIgnoredTokens: { [chainId: Hex]: { [key: string]: string[] } };
  allDetectedTokens: { [chainId: Hex]: { [key: string]: Token[] } };
};
```
i pushed a commit for it